packages/radix-ui-themes/src/components/tab-nav.tsx

```diff
@@ -46,27 +46,26 @@ interface TabNavLinkProps
   extends Omit<ComponentPropsWithoutColor<typeof NavigationMenu.Link>, 'onSelect'>,
     TabNavLinkOwnProps {}
 const TabNavLink = React.forwardRef<TabNavLinkElement, TabNavLinkProps>((props, forwardedRef) => {
-  const { asChild, className, children: childrenProp, ...linkProps } = props;
-
-  const { Root: TabNavLinkRoot, children } = getRoot({
-    asChild,
-    children: childrenProp,
-    parent: NavigationMenu.Link,
-  });
+  const { asChild, children, className, ...linkProps } = props;
 
   return (
     <NavigationMenu.Item className="rt-TabNavItem">
-      <TabNavLinkRoot
+      <NavigationMenu.Link
         {...linkProps}
         ref={forwardedRef}
         className={classNames('rt-reset', 'rt-BaseTabListTrigger', 'rt-TabNavLink', className)}
         onSelect={() => {}}
+        asChild={asChild}
       >
-        <span className="rt-BaseTabListTriggerInner rt-TabNavLinkInner">{children}</span>
-        <span className="rt-BaseTabListTriggerInnerHidden rt-TabNavLinkInnerHidden">
-          {children}
-        </span>
-      </TabNavLinkRoot>
+        {firstChildMightAdoptSubtree({ asChild, children }, (children) => (
```

Comment:
I went with the render prop approach as I felt it was actually the simplest to understand the overall abstraction in one go whilst still retaining good ergonomics looking at the tree and colocation of the inner content.

The alternative I tested was to have 2 functions:
- one placed where `firstChildMightAdoptSubtree` currently is but taking a "node" as second argument…
- …in which you'd call a second like `getConsumerChildren({ asChild, children })` inline in your nested DOM

Would look something like this for this component:

```tsx
<NavigationMenu.Link
  {...linkProps}
  ref={forwardedRef}
  className={classNames('rt-reset', 'rt-BaseTabListTrigger', 'rt-TabNavLink', className)}
  onSelect={() => {}}
  asChild={asChild}
>
  {firstChildMightAdoptSubtree({ asChild, children },
    <>
      <span className="rt-BaseTabListTriggerInner rt-TabNavLinkInner">
        {getConsumerChildren({ asChild, children })}
      </span>
      <span className="rt-BaseTabListTriggerInnerHidden rt-TabNavLinkInnerHidden">
        {getConsumerChildren({ asChild, children })}
      </span>
    </>
  )}
</NavigationMenu.Link>
```

Note that this case needs 2 instances of the children so when I had that, I actually prefered abstracting the inner stuff in its own component so `getConsumerChildren` was only called once.

Overall, I think the render prop does a better job but let me know what you think!